File: c_modules/printer_queue.c
#include <stdio.h>
#include <string.h>
#include "../modules/printer_queue.h"
#include "../modules/utils.h"

void add_print_job(const char user[])
{
    FILE *fp;
    char file[100], logmsg[200];

    printf("Enter file name to print: ");
    scanf("%s", file);

    fp = fopen("data/printer_queue.txt", "a");

    if(fp == NULL)
    {
        printf("Error opening queue file\n");
        return;
    }

    fprintf(fp,"%s %s\n", user, file);

    fclose(fp);

    printf("Job added to queue\n");

    sprintf(logmsg, "%s added print job: %s", user, file);
    log_activity(logmsg);
}

void process_print_job()
{
    FILE *fp;
    FILE *temp;

    char user[50], file[100], logmsg[200];
    int first = 1;

    fp = fopen("data/printer_queue.txt","r");

    if(fp == NULL)
    {
        printf("Queue empty\n");
        return;
    }

    temp = fopen("data/temp.txt","w");

    while(fscanf(fp,"%s %s", user, file) != EOF)
    {
        if(first)
        {
            printf("Printing: %s for %s\n", file, user);
            first = 0;
            sprintf(logmsg, "%s printed file: %s", user, file);
            log_activity(logmsg);
        }
        else
        {
            fprintf(temp,"%s %s\n", user, file);
        }
    }

    fclose(fp);
    fclose(temp);

    remove("data/printer_queue.txt");
    rename("data/temp.txt","data/printer_queue.txt");
}

void view_queue()
{
    clear_screen();
    FILE *fp;

    char user[50], file[100];
    int pos = 1;

    fp = fopen("data/printer_queue.txt","r");

    if(fp == NULL)
    {
        printf("Queue empty\n");
        return;
    }

    printf("\n===== Printer Queue (FCFS) =====\n");
    printf("Position\tUser\tFile\n");
    printf("----------------------------------\n");

    while(fscanf(fp,"%s %s", user, file) != EOF)
    {
        printf("%d\t\t%s\t%s\n", pos++, user, file);
    }

    fclose(fp);
}

void printer_menu(const char user[])
{
    int choice;

    while(1)
    {
        clear_screen();
        printf("\n===== Printer Menu =====\n");
        printf("1. Add Print Job\n");
        printf("2. Process Next Job\n");
        printf("3. View Queue\n");
        printf("4. Back to Main Menu\n");

        printf("Choice: ");
        scanf("%d",&choice);

        switch(choice)
        {
            case 1:
                add_print_job(user);
                pause_screen();
                break;

            case 2:
                process_print_job();
                pause_screen();
                break;

            case 3:
                view_queue();
                pause_screen();
                break;

            case 4:
                return;

            default:
                printf("Invalid choice\n");
        }
    }
}

File: modules/printer_queue.py
import time
import os
from modules.logs import log_activity

queue = []
printer_available = True

def clear():
    os.system('cls' if os.name == "nt" else "clear")

def printer_menu(user):

    while True:

        clear()
        print("\n===== Printer Menu =====")
        print("1. Add Print Job")
        print("2. Process Next Job")
        print("3. View Queue")
        print("4. Visualize Queue")
        print("5. Back to Main Menu")

        choice = input("Enter choice: ")

        if choice == "1":
            file = input("Enter file name to print: ")
            queue.append((user, file))
            print("Job added to queue")
            input("\nPress Enter to continue...")

        elif choice == "2":
            if len(queue) == 0:
                print("No print jobs in queue")
            else:
                job = queue.pop(0)
                print("Printing:", job[1], "for", job[0])
                print("Printing", end="")
                for i in range(5):
                    print(".", end="", flush=True)
                    time.sleep(0.5)
                
                print("\nPrint completed!\n")

            input("\nPress Enter to continue...")

        elif choice == "3":
            print("\nCurrent Queue:")
            if len(queue) == 0:
                print("Queue is empty")

            else:
                for i, job in enumerate(queue, start=1):
                    print(i, "-", job[0], ":", job[1])
            
            input("\nPress Enter to continue...")

        elif choice == "4":
            show_queue_visual()
            input("\nPress Enter to continue...")

        elif choice == "5":
            print("Returning to Main Menu...")
            break   

        else:
            print("Invalid choice")
            input("\nPress Enter to continue...")
    
def show_queue_visual():

    if len(queue) == 0:
        print("\nPrinter Queue Empty\n")
        return

    print("\nPrinter Queue\n")

    for job in queue:
        print(f"[ {job[1]} ]", end=" -> ")

    print("END\n")